crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -932,12 +935,59 @@ impl IndexerReader {
                 );
                 (inner_query, "1 = 1".into())
             }
-            Some(
-                TransactionFilter::TransactionKind(_) | TransactionFilter::TransactionKindIn(_),
-            ) => {
-                return Err(IndexerError::NotSupported(
-                    "TransactionKind filter is not supported.".into(),
-                ));
+            Some(TransactionFilter::TransactionKind(kind)) => {
+                if kind == IotaTransactionKind::SystemTransaction {
+                    ("tx_kinds".into(), "tx_kind = 0 OR tx_kind > 1".to_string())
+                } else {
+                    ("tx_kinds".into(), format!("tx_kind = {}", kind as u8))
+                }
+            }
+            Some(TransactionFilter::TransactionKindIn(kind_vec)) => {
+                if kind_vec.is_empty() {
+                    return Err(IndexerError::InvalidArgument(
+                        "no transaction kind provided".into(),
+                    ));
+                }
+
+                let mut has_system_transaction = false;
+                let mut has_programmable_transaction = false;
+                let mut other_kinds = Vec::new();
+
+                for kind in kind_vec.iter() {
+                    match kind {
+                        IotaTransactionKind::SystemTransaction => has_system_transaction = true,
+                        IotaTransactionKind::ProgrammableTransaction => {
+                            has_programmable_transaction = true
+                        }
+                        _ => other_kinds.push(*kind as u8),
```

Comment:
to add clarity if you agree we could make explicit to mark as `other` all not matched variants which are being pushed into this vector.
```suggestion
                        other => other_kinds.push(*other as u8),
```

Reply:
Yes, better let's use the variable: https://github.com/iotaledger/iota/pull/5801/commits/9e31a7bdb5047d1513b53116e88e40ec027d6b30